Remove _all_ codegen dependencies on `rustc_mir` :tada:

mark-i-m opened the pull request:
~This code is pretty self-contained. It has no references to the rest of `rustc_mir`. Moving it to its own crate means that almost all of the references from `rustc_codegen_*` to `rustc_mir` are instead moved to `rustc_monomorphize`, which should help improve compile times for the compiler a bit...~

With the help of eddyb and oli-obk, all of the dependencies of `librustc_codegen_*` on `librustc_mir` have been removed:
- dependencies on `rustc_mir::monomorphize` were moved to `rustc::mir::mono`
- `rustc_mir::const_eval::const_field` is made into a query.
- `rustc_mir::interpret::type_name` is made into a query.

This should help reduce compile time when working on `rustc_mir` :clock1: 

cc #47849

r? @eddyb

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/202360850) ([raw log](https://api.travis-ci.com/v3/job/202360850/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:1d07b310:start=1558571992361450311,finish=1558571993165491254,duration=804040943
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---
[00:06:59]    Compiling rustc_macros v0.1.0 (/checkout/src/librustc_macros)
[00:07:06]    Compiling syntax_pos v0.0.0 (/checkout/src/libsyntax_pos)
[00:07:10]    Compiling rustc_errors v0.0.0 (/checkout/src/librustc_errors)
[00:08:14]    Compiling syntax_ext v0.0.0 (/checkout/src/libsyntax_ext)
[00:14:22]    Compiling rustc_monomorphize v0.0.0 (/checkout/src/librustc_monomorphize)
[00:17:25]    Compiling rustc_allocator v0.0.0 (/checkout/src/librustc_allocator)
[00:17:26]    Compiling rustc_lint v0.0.0 (/checkout/src/librustc_lint)
[00:18:21]    Compiling rustc_traits v0.0.0 (/checkout/src/librustc_traits)
[00:18:23]    Compiling rustc_mir v0.0.0 (/checkout/src/librustc_mir)
---
[00:27:27]    Compiling rustc_lsan v0.0.0 (/checkout/src/librustc_lsan)
[00:27:27]    Compiling rustc_msan v0.0.0 (/checkout/src/librustc_msan)
[00:27:28]    Compiling rustc_tsan v0.0.0 (/checkout/src/librustc_tsan)
[00:27:28]    Compiling rustc_asan v0.0.0 (/checkout/src/librustc_asan)
[00:27:53] error: internal compiler error: src/librustc/ty/query/mod.rs:101: tcx.collect_and_partition_mono_items(crate0) unsupported by its crate
[00:27:53] thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:637:9
[00:27:53] note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
[00:27:53] error: aborting due to previous error
[00:27:53] 
[00:27:53] 
[00:27:53] 
[00:27:53] note: the compiler unexpectedly panicked. this is a bug.
[00:27:53] 
[00:27:53] note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
[00:27:53] 
[00:27:53] note: rustc 1.36.0-dev running on x86_64-unknown-linux-gnu
[00:27:53] 
[00:27:53] note: compiler flags: -Z external-macro-backtrace -Z unstable-options -Z force-unstable-if-unmarked -C opt-level=2 -C prefer-dynamic -C debug-assertions=y -C codegen-units=1 -C link-args=-Wl,-rpath,$ORIGIN/../lib --crate-type lib
[00:27:53] note: some of the compiler flags provided by cargo are hidden
[00:27:53] 
[00:27:53] error: Could not compile `core`.
[00:27:53] 
---
19656 ./obj/build/x86_64-unknown-linux-gnu/stage0-std/release
19256 ./src/llvm-project/lldb/www/cpp_reference
19252 ./src/llvm-project/lldb/www/cpp_reference/html
travis_time:end:1664d2d0:start=1558573677533508712,finish=1558573678169073130,dura[0K$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:00d1a2ee
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

mark-i-m commented:
All tests passing locally now :)

eddyb commented on src/librustc_codegen_llvm/debuginfo/namespace.rs:
```diff
@@ -11,6 +10,7 @@ use crate::common::CodegenCx;
 use rustc::hir::def_id::DefId;
 use rustc::hir::map::DefPathData;
 
+use rustc_monomorphize::Instance;
```
Isn't `Instance` a reexport of something from `rustc`? We should probably refer to the original in `rustc` noawadays.

mark-i-m replied:
Ah, yes. It is a re-export of `rustc::ty::Instance`. Should I just remove the re-export altogether?

mark-i-m replied:
I went ahead and removed it. Let me know what you think.

eddyb commented on src/librustc_mir/interpret/traits.rs:
```diff
@@ -55,7 +55,7 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tcx, M>
         let tcx = &*self.tcx;
 
-        let drop = crate::monomorphize::resolve_drop_in_place(*tcx, ty);
+        let drop = rustc_monomorphize::resolve_drop_in_place(*tcx, ty);
```
This should be a query.

eddyb replied:
And/or in `rustc` itself.

eddyb commented on src/librustc_mir/interpret/traits.rs:
```diff
@@ -55,7 +55,7 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tcx, M>
         let tcx = &*self.tcx;
 
-        let drop = crate::monomorphize::resolve_drop_in_place(*tcx, ty);
+        let drop = rustc_monomorphize::resolve_drop_in_place(*tcx, ty);
```
This, and other `Instance` constructors, should be moved to `Instance` methods in `rustc`, instead.

mark-i-m replied:
It appears there is _a lot_ of repeated code, so I deduplicated it and move it to methods on `rustc::ty::Instance`

eddyb commented on src/librustc_monomorphize/lib.rs:
```diff
@@ -0,0 +1,90 @@
+    }
+}
+
+pub fn custom_coerce_unsize_info<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
```
cc @oli-obk Surely this could be somewhere that miri would also benefit from?

oli-obk replied:
since `librustc_monomorphize` is a dependency of `librustc_mir`, that should be easy enough

eddyb commented on src/librustc_monomorphize/lib.rs:
```diff
@@ -0,0 +1,90 @@
+
+#[inline(never)] // give this a place in the profiler
+pub fn assert_symbols_are_distinct<'a, 'tcx, I>(tcx: TyCtxt<'a, 'tcx, 'tcx>, mono_items: I)
+    where I: Iterator<Item=&'a MonoItem<'tcx>>
```
This looks like it should be in `rustc_codegen_utils::symbol_names` or somewhere similar?

mark-i-m replied:
Moved it, as mentioned in your other comment.

eddyb commented on src/librustc_monomorphize/partitioning.rs:
```diff
@@ -922,7 +925,7 @@ fn collect_and_partition_mono_items<'a, 'tcx>(
     tcx.sess.abort_if_errors();
 
-    crate::monomorphize::assert_symbols_are_distinct(tcx, items.iter());
+    crate::assert_symbols_are_distinct(tcx, items.iter());
```
Wait, is this the only call to this function? Can't it be moved into this module then? (And not made public?)

mark-i-m replied:
To a large extent, I mostly just moved files. I didn't really do any deeper cleanup.

I'll move the function :+1:

eddyb commented on src/librustc_monomorphize/partitioning.rs:
```diff
@@ -110,9 +113,9 @@ use rustc::util::common::time;
+use crate::collector::{self, MonoItemCollectionMode};
+use crate::item::{MonoItemExt, InstantiationMode};
 
 pub use rustc::mir::mono::CodegenUnit;
```
This shouldn't be a reexport anymore.

eddyb commented on src/librustc_monomorphize/item.rs:
```diff
@@ -1,5 +1,8 @@
-use crate::monomorphize::Instance;
-use rustc::hir;
+pub use rustc::mir::mono::MonoItem;
```
This shouldn't be a reexport anymore.

eddyb commented:
Looks like the non-reexport items are mostly extension traits. We should figure out a way to break the dependency without necessarily moving the code out of `rustc_mir`.

mark-i-m commented:
@eddyb I think I addressed most of your review concerns.

> Looks like the non-reexport items are mostly extension traits. We should figure out a way to break the dependency without necessarily moving the code out of rustc_mir.

I thought about this a bit. It might also be possible to move them to `librustc` directly, if that seems appropriate. That is, I don't think there are any cyclic dependencies.

mark-i-m commented:
In fact, that may also enable moving parts of `librustc_codegen_utils` to `librustc` too, but I haven't checked.

eddyb commented on src/librustc_codegen_llvm/type_of.rs:
```diff
@@ -4,7 +4,7 @@ use crate::type_::Type;
 use rustc::ty::layout::{self, Align, LayoutOf, FnTypeExt, PointeeInfo, Size, TyLayout};
 use rustc_target::abi::{FloatTy, TyLayoutMethods};
-use rustc_mir::monomorphize::item::DefPathBasedNames;
+use rustc_monomorphize::item::DefPathBasedNames;
```
Okay, I looked at this, and we can probably move `DefPathBasedNames` fully to `ty::print`, in a submodule named `obsolete`, and open an issue about refactoring it to use `ty::print` infra (I already left comments, I believe, but they do not have an issue attached).

eddyb replied:
Oh and once we do that, `MonoItemExt` can just be inherent methods on `MonoItem`.
Also, I'm a bit surprised `MonoItem::GlobalAsm` uses a `HirId` and not a `DefId`, that's weird.

mark-i-m replied:
@eddyb It looks like `librustc_codegen_ssa` also uses `MonoItemExt` :(

https://github.com/rust-lang/rust/blob/27cc0db7a248308fc2634ac68d7608a20b4a1c09/src/librustc_codegen_ssa/mono_item.rs#L10-L12

Moving those methods to `rustc::mir::mono` is hard because they use `trait rustc_codegen_ssa::traits::BuilderMethods`, which itself uses half a dozen other traits from the that crate.

We could try to move the transitive closure of all of those traits to `librustc`, but things could get messy... I don't know if it will work.

I have yet to investigate `CodegenUnitExt`...

eddyb replied:
I am talking about `rustc_mir::monomorphize::item::MonoItemExt`, not `rustc_codegen_ssa::mono_item::MonoItemExt` - you can ignore the latter, it's part of codegen extensions to non-codegen types.

mark-i-m replied:
@eddyb `rustc_codegen_ssa::mono_item::MonoItemExt: rustc_mir::monomorphize::item::MonoItemExt`

eddyb replied:
That's only for the trick of being able to put the body of the methods in the trait, as default methods, by relying on `as_mono_item`, the methods will still obviously be available on `MonoItem`, they're not going away, so the trait isn't needed.

mark-i-m replied:
Heh, sorry my brain was missing in action.

eddyb commented:
@mark-i-m I quickly looked at it, and both `MonoItemExt` and `CodegenUnitExt` should be movable to `rustc::mir::mono` as inherent methods.

mark-i-m commented:
@eddyb yes, but that trait uses this one.

eddyb commented:
@mark-i-m Yes, but not in a way which needs `rustc_codegen_ssa::mono_item::MonoItemExt` to be moved, in order to move `rustc_mir::monomorphize::item::MonoItemExt` to inherent methods.

mark-i-m commented:
@eddyb Done. `librustc_codegen_llvm` now no longer depends on `librustc_mir` _or_ `librustc_monomorphize` :tada:

eddyb commented on src/librustc_codegen_ssa/mono_item.rs:
```diff
@@ -87,4 +86,8 @@ pub trait MonoItemExt<'a, 'tcx: 'a>: fmt::Debug + BaseMonoItemExt<'a, 'tcx> {
+impl<'a, 'tcx: 'a> MonoItemExt<'a, 'tcx> for MonoItem<'tcx> {
+    fn as_mono_item(&self) -> &MonoItem<'tcx> {
+        self
+    }
```
If you move the bodies of the methods here you don't even need `as_mono_item` or calls to it.

eddyb commented on src/librustc_codegen_ssa/Cargo.toml:
```diff
@@ -34,4 +34,5 @@ rustc_errors = { path = "../librustc_errors" }
 rustc_fs_util = { path = "../librustc_fs_util" }
 rustc_incremental = { path = "../librustc_incremental" }
 rustc_mir = { path = "../librustc_mir" }
+rustc_monomorphize = { path = "../librustc_monomorphize" }
```
Is this needed?

mark-i-m replied:
Not anymore :tada: Good catch!

eddyb commented on src/librustc_codegen_utils/Cargo.toml:
```diff
@@ -20,4 +20,4 @@ rustc = { path = "../librustc" }
 rustc_data_structures = { path = "../librustc_data_structures" }
 rustc_metadata = { path = "../librustc_metadata" }
-rustc_mir = { path = "../librustc_mir" }
+rustc_monomorphize = { path = "../librustc_monomorphize" }
```
Is this needed?

mark-i-m replied:
Likewise :tada:

eddyb commented on src/librustc_mir/Cargo.toml:
```diff
@@ -20,6 +20,7 @@ rustc = { path = "../librustc" }
 rustc_target = { path = "../librustc_target" }
 rustc_data_structures = { path = "../librustc_data_structures" }
 rustc_errors = { path = "../librustc_errors" }
+rustc_monomorphize = { path = "../librustc_monomorphize" }
```
Is this needed?

mark-i-m replied:
Same here.

mark-i-m commented:
@eddyb Yes, I will move monomorphize back to rustc_mir.

However, there is still one there is still one dependency from `librustc_codegen_ssa` to `librustc_mir`: `rustc_mir::const_eval::const_field`. However, it looks like that might also be movable to `librustc` (not sure though).

mark-i-m commented:
`const_field` is harder to move because it uses a lot of the const eval machinery in `rustc_mir::interpret`. @oli-obk any suggestions?

eddyb commented:
r? @oli-obk for `rustc_mir::const_eval::const_field`

mark-i-m commented:
@oli-obk Just from some greping, it looks like it might be possible to just move `rustc_mir::interpret` to `rustc::interpret` altogether. Does this sound right to you? Should I do it if possible?

eddyb commented:
>  just move `rustc_mir::interpret` to `rustc::interpret`

IMO that's a bad idea, because the former is *the entirety of miri*, which is not small.

When you can't move something because it has a lot of dependencies, I'd suggest trying to put it behind a query, which does break dependencies (as long as the signature doesn't require more context than a `tcx`).

mark-i-m commented:
@eddyb This is the definition of `const_field`. IIUC, this should be easily query-izable (all of the types are in `librustc`, and it takes a `tcx` already), right? If so, I can try that later today.

```rust
pub fn const_field<'a, 'tcx>(
    tcx: TyCtxt<'a, 'tcx, 'tcx>,
    param_env: ty::ParamEnv<'tcx>,
    variant: Option<VariantIdx>,
    field: mir::Field,
    value: ty::Const<'tcx>,
) -> ty::Const<'tcx> 
```

oli-obk commented:
>  I see that queries can only have one key. How is this problem normally solved?

Either by using a tuple or by creating a new struct that contains all info

mark-i-m commented:
@oli-obk It seems that `std::option::Option<rustc_target::abi::VariantIdx>` and `mir::Field` do not `impl ty::query::keys::Key`. Do I just use this?

```rust
impl<'tcx> Key for (Option<VariantIdx>, mir::Field, ty::Const<'tcx>) {
    fn query_crate(&self) -> CrateNum {
        LOCAL_CRATE
    }
    fn default_span(&self, _: TyCtxt<'_, '_, '_>) -> Span {
        DUMMY_SP
    }
}

```

mark-i-m commented:
And likewise: 

the trait bound `ty::ParamEnvAnd<'_, (std::option::Option<rustc_target::abi::VariantIdx>, mir::Field, ty::sty::Const<'_>)>: dep_graph::dep_node::RecoverKey<'_>` is not satisfied

mark-i-m commented:
@oli-obk Can you sanity-check this commit: https://github.com/rust-lang/rust/pull/61062/commits/b89711d72b1854842723f4a027c76b1dbfa48642

oli-obk commented:
r=eddyb,oli-obk with the `const_field` nit + travis

mark-i-m commented:
Travis passed :)

eddyb commented on src/librustc/query/mod.rs:
```diff
@@ -431,6 +432,15 @@ rustc_queries! {
             }
         }
+
+        /// Projects to a field of a (variant of a) const.
```
Since this is about values and not places, maybe "Extract" is a better verb?

eddyb commented on src/librustc/query/mod.rs:
```diff
@@ -431,6 +432,15 @@ rustc_queries! {
+
+        /// Projects to a field of a (variant of a) const.
+        query const_field(
+            key: ty::ParamEnvAnd<'tcx, (mir::Field, ty::Const<'tcx>)>
```
I would put the `Const` before `Field`, to make it a bit clearer.

eddyb commented on src/librustc/ty/print/mod.rs:
```diff
@@ -7,6 +7,7 @@ use rustc_data_structures::fx::FxHashSet;
 // `pretty` is a separate module only for organization.
 mod pretty;
+pub mod obsolete;
 pub use self::pretty::*;
```
Hmm, maybe put `mod obsolete; ` below the `pretty` reexport (with an empty line before it).

eddyb commented on src/librustc_codegen_ssa/mir/constant.rs:
```diff
@@ -46,13 +45,13 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
-                        mir::Field::new(field as usize),
-                        c,
-                    );
+                    let field = bx.tcx().const_field(ty::ParamEnvAnd {
```
I think this should be using `.and` on the `ParamEnv`.

eddyb commented on src/librustc_mir/lib.rs:
```diff
@@ -51,19 +51,22 @@ mod shim;
     providers.const_eval = const_eval::const_eval_provider;
     providers.const_eval_raw = const_eval::const_eval_raw_provider;
     providers.check_match = hair::pattern::check_match;
+    providers.const_field = |tcx, ParamEnvAnd { param_env, value: (field, value) }| {
```
Do we do this anywhere else? I could've sworn it looked different but now I'm not sure where to tell you to look for an example (`rustc::ty::layout` maybe?)

mark-i-m replied:
It looks like it's done in a lot of places, but usually the key isn't destructured. I can move the destructuring to a `let` binding in the body if you like.

```console
$ rg --iglob '*.rs' 'providers.*\|tcx'
src/librustc_codegen_llvm/attributes.rs
317:    providers.target_features_whitelist = |tcx, cnum| {
337:    providers.wasm_import_module_map = |tcx, cnum| {

src/librustc_codegen_ssa/base.rs
860:    providers.backend_optimization_level = |tcx, cratenum| {
893:    providers.dllimport_foreign_items = |tcx, krate| {
918:    providers.is_dllimport_foreign_item = |tcx, def_id| {

src/librustc_mir/monomorphize/partitioning.rs
969:    providers.is_codegened_item = |tcx, def_id| {
975:    providers.codegen_unit = |tcx, name| {

src/librustc/ty/context.rs
3048:    providers.in_scope_traits_map = |tcx, id| tcx.gcx.trait_map.get(&id);
3049:    providers.module_exports = |tcx, id| tcx.gcx.export_map.get(&id).map(|v| &v[..]);
3050:    providers.crate_name = |tcx, id| {
3054:    providers.get_lib_features = |tcx, id| {
3058:    providers.get_lang_items = |tcx, id| {
3062:    providers.upvars = |tcx, id| tcx.gcx.upvars.get(&id).map(|v| &v[..]);
3063:    providers.maybe_unused_trait_import = |tcx, id| {
3066:    providers.maybe_unused_extern_crates = |tcx, cnum| {
3070:    providers.names_imported_by_glob_use = |tcx, id| {
3075:    providers.stability_index = |tcx, cnum| {
3079:    providers.lookup_stability = |tcx, id| {
3084:    providers.lookup_deprecation_entry = |tcx, id| {
3089:    providers.extern_mod_stmt_cnum = |tcx, id| {
3093:    providers.all_crate_nums = |tcx, cnum| {
3097:    providers.postorder_cnums = |tcx, cnum| {
3101:    providers.output_filenames = |tcx, cnum| {
3105:    providers.features_query = |tcx, cnum| {
3109:    providers.is_panic_runtime = |tcx, cnum| {
3113:    providers.is_compiler_builtins = |tcx, cnum| {

src/test/run-make-fulldeps/hotplug_codegen_backend/the_backend.rs
50:        providers.target_features_whitelist = |tcx, _cnum| {

src/librustc_typeck/check/method/suggest.rs
844:    providers.all_traits = |tcx, cnum| {
```

eddyb replied:
Sorry, I was referring to the `ParamEnvAnd` destructuring, since I could've sworn I've never seen it used.

eddyb replied:
Yeah, this is how we actually do this: https://github.com/rust-lang/rust/blob/fe5f42cdb88d8ce31f746130099321e7c95e1ef0/src/librustc/ty/util.rs#L940-L944

bors commented:
:umbrella: The latest upstream changes (presumably #59276) made this pull request unmergeable. Please resolve the merge conflicts.

mark-i-m commented:
@eddyb I think I addressed your review comments.

@oli-obk I rebased (and try to clean up the git history a bit by squashing commits, but it is still a bit messy)

oli-obk commented:
r? @eddyb for the destructuring style (not clear what you mean by "this" in "do we do this anywhere else?")

bors commented:
:umbrella: The latest upstream changes (presumably #60955) made this pull request unmergeable. Please resolve the merge conflicts.

RalfJung commented on src/librustc_mir/const_eval.rs:
```diff
@@ -468,7 +468,7 @@ impl<'a, 'mir, 'tcx> interpret::Machine<'a, 'mir, 'tcx>
+/// Extracts a field of a (variant of a) const.
 // this function uses `unwrap` copiously, because an already validated constant must have valid
 // fields and can thus never fail outside of compiler bugs
 pub fn const_field<'a, 'tcx>(
```
Our other queries here are called `*_provider`. Might make sense to be consistent about that?

eddyb replied:
Using the name of the query is proper, IMO - however, it should definitely not be public!

mark-i-m replied:
It is public because it is called directly elsewhere, as per @oli-obk... can’t find the comment atm.

mark-i-m replied:
@eddyb ^^^

eddyb commented:
@mark-i-m Can you rebase, and fix https://github.com/rust-lang/rust/pull/61062#discussion_r288150002?
r=me after that

mark-i-m commented:
@eddyb Rebased and addressed nit. Unfortunately, someone added another dependency on rustc_mir in the meantime (on `rustc_mir::interpret::type_name`), so I turned that into a query too in the last commit. Let me know what you think.

mark-i-m commented:
Also, I'm glad to address nits, but if we could do it in a followup PR, that would be great, as this rebase was painful...

RalfJung commented:
> someone added another dependency on rustc_mir in the meantime (on rustc_mir::interpret::type_name), so I turned that into a query too in the last commit. Let me know what you think.

Cc @oli-obk

oli-obk commented:
@bors r=eddyb,oli-obk p=1

`type_name` lgtm, prioritizing due to bitrottiness

bors commented:
:pushpin: Commit 0f822d775f046df87713edf8ffb4e308fc77b3ef has been approved by `eddyb,oli-obk`

<!-- @bors r=eddyb,oli-obk 0f822d775f046df87713edf8ffb4e308fc77b3ef -->
<!-- homu: {"type":"Approved","sha":"0f822d775f046df87713edf8ffb4e308fc77b3ef","approver":"eddyb,oli-obk"} -->

bors commented:
:hourglass: Testing commit 0f822d775f046df87713edf8ffb4e308fc77b3ef with merge da9ebc828c982d2ed49396886da85011e1b0a6c0...
<!-- homu: {"type":"BuildStarted","head_sha":"0f822d775f046df87713edf8ffb4e308fc77b3ef","merge_sha":"da9ebc828c982d2ed49396886da85011e1b0a6c0"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/114062724), [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/24996042)
Approved by: eddyb,oli-obk
Pushing da9ebc828c982d2ed49396886da85011e1b0a6c0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"eddyb,oli-obk","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang/rust/builds/24996042","checks-travis":"https://travis-ci.com/rust-lang/rust/builds/114062724"},"merge_sha":"da9ebc828c982d2ed49396886da85011e1b0a6c0"} -->

eddyb commented:
@oli-obk We should probably put miri intrinsics behind `const_eval` with the `DefId` of the intrinsic (effectively treating a 0-argument call as a use of a constant).

oli-obk commented:
so... const fns with zero arguments and constants become indistinguishable?
